File: 2- Elementary Sorts/Cocktail Sort/Cocktail.cpp
#include <vector>
#include <cstdlib>
#include <ctime>
using namespace std;
class Cocktail {
private:
    template<typename T>
    static void swap(T *xp, T *yp){
        T temp = *xp;
        *xp = *yp;
        *yp = temp;
    }
public:
    template<typename T>
    static void Sort(vector<T> &a) {
        bool swapped = true;
        int start = 0;
        int n = a.size();
        int end = n - 1;
        while (swapped){
            swapped = false;
            for (int i = start; i < end; ++i){
                if (a[i] > a[i + 1]) {
                    swap(&a[i], &a[i + 1]);
                    swapped = true;
                }
            }
            if (!swapped)
                break;
            swapped = false;
            --end;
            for (int i = end - 1; i >= start; --i){
                if (a[i] > a[i + 1]) {
                    swap(&a[i], &a[i + 1]);
                    swapped = true;
                }
            }
            ++start;
        }
    }
};

File: 2- Elementary Sorts/Cocktail Sort/Cocktail.java
import java.util.ArrayList;

public class Cocktail {
    public static <T extends Comparable<T>> void Sort(ArrayList<T> a) {
        boolean swapped = true;
        int start = 0;
        int end = a.size();
        while (swapped == true) {
            swapped = false;
            for (int i = start; i < end - 1; ++i) {
                if (a.get(i).compareTo(a.get(i+1)) > 0) {
                    T temp = a.get(i);
                    a.set(i,a.get(i+1));
                    a.set(i+1,temp);
                    swapped = true;
                }
            }
            if (swapped == false)
                break;
            swapped = false;
            end = end - 1;
            for (int i = end - 1; i >= start; i--) {
                if (a.get(i).compareTo(a.get(i+1)) > 0) {
                    T temp = a.get(i);
                    a.set(i,a.get(i+1));
                    a.set(i+1,temp);
                    swapped = true;
                }
            }
            start = start + 1;
        }
    }
}
